### PR TITLE
MP3 code Withdrawn-- code from Arthur Taylor is better

### PR DESCRIPTION
**Pull request by jpffitch**
Think this does it.  Not implemented writing by frames but that is really just copying code.  Would appreciate any feedback

### CODE REVIEW

File: src/mp3.c [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+#include	<string.h>
+#include	<ctype.h>
+#include	<math.h>
+
+#include	"sndfile.h"
+#include	"sfendian.h"
+#include	"common.h"
+#include	"mp3.h"
+
+#if 0
```
Comment (erikd) [inline src/mp3.c]: We can just remove that `#if 0`.

File: src/sndfile.c [hunk 2]
```diff
@@ -8,7 +8,7 @@
 **
 ** This program is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
-** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
```
Comment (erikd) [inline src/sndfile.c]: ??

File: src/sndfile.c [hunk 3]
```diff
@@ -59,7 +59,7 @@ ErrorStruct SndfileErrors [] =
 	/* Public error values and their associated strings. */
 	{	SF_ERR_NO_ERROR				, "No Error." },
 	{	SF_ERR_UNRECOGNISED_FORMAT	, "Format not recognised." },
-	{	SF_ERR_SYSTEM				, "System error." /* Often replaced. */ 	},
+	{	SF_ERR_SYSTEM				, "System error." /* Often replaced. */		},
```
Comment (erikd) [inline src/sndfile.c]: ??

File: src/sndfile.c [hunk 4]
```diff
@@ -325,7 +325,7 @@ static char	sf_syserr [SF_SYSERR_LEN] = { 0 } ;
 
 SNDFILE*
 sf_open	(const char *path, int mode, SF_INFO *sfinfo)
-{	SF_PRIVATE 	*psf ;
+{	SF_PRIVATE	*psf ;
```
Comment (erikd) [inline src/sndfile.c]: Seems to be a whole bunch of semantically null changes in this PR. They should be removed.

File: src/sndfile.h.in [hunk 5]
```diff
@@ -116,7 +116,14 @@ enum
...
 
-	/* Endian-ness options. */
+	SF_FORMAT_MP3_STEREO		= 0x0080,		/* MP3 MODE. */
+	SF_FORMAT_MP3_JTSTEREO		= 0x0081,		/* MP3 MODE. */
+	SF_FORMAT_MP3_DUAL		= 0x0082,		/* MP3 MODE. */
+	SF_FORMAT_MP3_MONO		= 0x0083,		/* MP3 MODE. */
+	/* Endian-nes
+
+
+ options. */
```
Comment (erikd) [inline src/sndfile.h.in]: The comment there should be unchanged.

File: src/mp3.c [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+	if ((mdata = calloc (1, sizeof (MP3_PRIVATE))) == NULL)
+		return SFE_MALLOC_FAILED ;
+	psf->container_data = mdata ;
+
+	if (psf->file.mode == SFM_RDWR)
+		return SFE_BAD_MODE_RW ;
+
+
+	if (psf->file.mode == SFM_WRITE)
+	{
```
Comment (erikd) [inline src/mp3.c]: No need for curly braces in this block.

File: src/mp3.c [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+
+		return mp3_open_lame (psf) ;
+	}
+
+	if (psf->file.mode == SFM_RDWR)
+		return SFE_UNIMPLEMENTED ;
+
+
+	if (psf->file.mode == SFM_READ)
+		return SFE_UNIMPLEMENTED ;
```
Comment (erikd) [inline src/mp3.c]: Will need to support read before this can be merged.

File: src/mp3.c [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+mp3_lame_parameters (lame_global_flags *gfp, MP3_PRIVATE *p)
+{	int error ;
+	lame_set_quality (gfp, p->quality) ;
+	if ((error = lame_init_params (gfp)) < 0)
+		return SFE_BAD_OPEN_FORMAT ;
+	p->initialised = 1 ;
+	return 1 ;
+}
+
+/* **** Note that if mode is MONO (3) then left and right channels are averaged and the data is changed.  Hence the _M functions **** */
```
Comment (erikd) [inline src/mp3.c]: This stereo mix down to mono functionality should probably be dropped.

File: src/mp3.c (hunk 1 above)
Comment (jpffitch) [reply src/mp3.c]: Done; thank you for all your coments.

File: src/sndfile.c (hunk 2 above)
Comment (jpffitch) [reply src/sndfile.c]: Artifact of using tabify in emacs; picked a space which would be the same as tab.  Fixed this and the others I believe.

File: src/sndfile.h.in (hunk 5 above)
Comment (jpffitch) [reply src/sndfile.h.in]: Sorry; my bad typing and I did not notice.  fixed.

File: src/mp3.c (hunk 6 above)
Comment (jpffitch) [reply src/mp3.c]: artifact of some earlier debugging code -- fixed

File: src/mp3.c (hunk 7 above)
Comment (jpffitch) [reply src/mp3.c]: Yes; that is my current focus.  I still have multiple routes and do not know which way to go.  I am still investigating.  mpg123 seems heavy and foolswood's attempt seems to have stopped.  But I have taken a copy of his code to investigate.  minimp3 is wacky, being all one include file.  May be the lightest way.   There is code in LAME based on mpg123 but the documention/examples I have found look sparse.

File: src/mp3.c (hunk 8 above)
Comment (jpffitch) [reply src/mp3.c]: So you suggest mono is signalled only by sf.channels?  That still leaves the two types of stereo to resolve.  I do need to ensure sf.channels is 1 or 2 only.

Comment (erikd) [reply src/mp3.c]: > So you suggest mono is signalled only by sf.channels? 

Yes.

> I do need to ensure sf.channels is 1 or 2 only.

A number of formats already require things like that. See `sf_format_check` in `src/sndfile.c`.

File: src/mp3.c [hunk 9]
```diff
@@ -407,23 +410,26 @@ static int mp3_close_read (SF_PRIVATE *psf)
 }
 
 #define CHUNK_SIZE	(512)
+#define FRAME_SIZE	(1152)
+#define NORMALISE	(0.000030517578125)
```
Comment (erikd) [inline src/mp3.c]: What is that magic number all about?

Comment (jpffitch) [reply src/mp3.c]: 1/32K to change short range to [-1, =1]

Comment (jpffitch) [reply src/mp3.c]: This code now removed.

File: src/mp3.h [hunk 10]
```diff
@@ -20,8 +20,8 @@
 
 #ifndef SF_SRC_MP3_H
 
-#include "lame.h"
-#include "mpg123.h"
+#include <lame/lame.h>
+#include <mpg123.h>
```
Comment (erikd) [inline src/mp3.h]: :100:

**Pull request closed without merging**